staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1864,3 +1986,523 @@ func TestForgetWatcher(t *testing.T) {
 	assertCacherInternalState(0, 0)
 	require.Equal(t, 2, forgetCounter)
 }
+
+// TestGetWatchCacheResourceVersion test the following cases:
+//
+// +-----------------+---------------------+-----------------------+
+// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+// +=================+=====================+=======================+
+// | Unset           | true/false          | nil/true/false        |
+// | 0               | true/false          | nil/true/false        |
+// | 105             | true/false          | nil/true/false        |
+// +-----------------+---------------------+-----------------------+
+// where:
+// - false indicates the value of the param was set to "false" by a test case
+// - true  indicates the value of the param was set to "true" by a test case
+func TestGetWatchCacheResourceVersion(t *testing.T) {
+	listOptions := func(allowBookmarks bool, sendInitialEvents *bool, rv string) storage.ListOptions {
+		p := storage.Everything
+		p.AllowWatchBookmarks = allowBookmarks
+
+		opts := storage.ListOptions{}
+		opts.Predicate = p
+		opts.SendInitialEvents = sendInitialEvents
+		opts.ResourceVersion = rv
+		return opts
+	}
+
+	scenarios := []struct {
+		name string
+		opts storage.ListOptions
+
+		expectedWatchResourceVersion int
+	}{
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | Unset           | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name: "RV=unset, allowWatchBookmarks=true, sendInitialEvents=nil, expectedRV=0",
+			opts: listOptions(true, nil, ""),
```

Comment:
This case seems wrong to me - if RV is unset, I would always expect the expectedRV to be the RV from storage...

Reply:
This really seems like this case:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L517

[which we will be able to remove after graduating to GA if we handle it correctly here]

Reply:
If we do that then I think we will change semantics for non streaming api requests. The intention of this PR was to change the semantics only for streaming APIs.

Reply:
if RV="" - that explicitly means user intention is "consistent read/watch"

Can you point me to place where we handle it differently - if there is such place that most probably means a bug.

Reply:
i'm lost, I though that `"consistent read/watch"` is when `SendInitialEvents == true` and `ResourceVersion = ""`